helix-term/src/commands/typed.rs

```diff
@@ -2167,6 +2169,64 @@ fn reset_diff_change(
     Ok(())
 }
 
+fn rename_buffer(
+    cx: &mut compositor::Context,
+    args: &[Cow<str>],
+    event: PromptEvent,
+) -> anyhow::Result<()> {
+    if event != PromptEvent::Validate {
+        return Ok(());
+    }
+    ensure!(args.len() == 1, ":rename takes one argument");
+    let new_name = args.first().unwrap();
+
+    let (_, doc) = current!(cx.editor);
+    let path = doc
+        .relative_path()
+        .ok_or_else(|| anyhow!("Buffer not saved, use :write"))?
+        .canonicalize()
+        .map_err(|_| anyhow!("Could not get absolute path of the document"))?;
+    let mut path_new = path.clone();
+    path_new.set_file_name(OsStr::new(new_name.as_ref()));
+    if path_new.exists() {
+        bail!("This file already exists");
+    }
+
+    if let Err(e) = std::fs::rename(&path, &path_new) {
+        bail!("Could not rename file, error: {}", e);
+    }
+    let (_, doc) = current!(cx.editor);
+    doc.set_path(Some(path_new.as_path()))
+        .map_err(|_| anyhow!("File renamed, but could not set path of the document"))?;
```

Comment:
i think that this should probably happen before the rename. If the fails later we can roll this back pretty easily (ideally with an unchecked method) while rolling back IO is pretty hard